src/compiler/checker.ts

```diff
@@ -6609,16 +6608,18 @@ namespace ts {
         }
 
         function getTypeReferenceType(node: TypeReferenceNode | ExpressionWithTypeArguments | JSDocTypeReference, symbol: Symbol) {
+            const typeArguments = typeArgumentsFromTypeReferenceNode(node); // Do unconditionally so we mark type arguments as referenced.

```

Comment:
Are you sure it's not cheap enough to just call `typeArgumentsFromTypeReferenceNode` multiple times and avoid the argument-threading churn below?